### PR TITLE
Separate TrainingModel into TrainingModel and EarlyStoppingTrainer classes.

### PR DESCRIPTION
**Pull request by fhieber**
- split TrainingModel class into TrainingModel and Trainer class to separate model from fitting logic better
- simplified training resumption logic
- TrainState stores more information about training state
- removed the quite complicated and convoluted callback logic
- restructured fit method in Trainer to make it more readable
- introduced OptimizerConfig to bundle optimizer-related settings
- CheckpointDecoder now returns overall decoding time, not only sent/sec.
- Fixed a bug with gradient norms not being written to tensorboard correctly
- Fixed the same bug as in #302 
- simplified constructors of {Sockeye,Training,Inference}Model

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./pre-commit.sh` or manual run of pylint & mypy)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/train.py [hunk 1]
```diff
@@ -174,26 +175,26 @@ def determine_context(args: argparse.Namespace, exit_stack: ExitStack) -> List[m
...
+                              train_context: List[mx.Context]) -> Optional[checkpoint_decoder.CheckpointDecoder]:
     """
-    Determine the number of sentences to decode and the context we should run on (CPU or GPU).
+    Returns a checkpoint decoder or None.
 
     :param args: Arguments as returned by argparse.
     :param exit_stack: An ExitStack from contextlib.
     :param train_context: Context for training.
-    :return: The number of sentences to decode and a list with the context(s) to run on.
+    :return: CheckpointDecoder instance or None.
```
[the remark this answers is not in the document]
Comment (fhieber) [reply sockeye/train.py]: done

File: sockeye/train.py [hunk 2]
```diff
@@ -724,24 +687,42 @@ def define_optimizer(args, lr_scheduler_instance) -> Tuple[str, Dict, str, str,
...
+    config = OptimizerConfig(name=args.optimizer,
+                             params=optimizer_params,
+                             kvstore=args.kvstore,
+                             initializer=initializer.get_initializer(
+                                 default_init_type=args.weight_init,
+                                 default_init_scale=args.weight_init_scale,
+                                 default_init_xavier_rand_type=args.weight_init_xavier_rand_type,
+                                 default_init_xavier_factor_type=args.weight_init_xavier_factor_type,
+                                 embed_init_type=args.embed_weight_init,
+                                 embed_init_sigma=source_vocab_sizes[0] ** -0.5,  # TODO
```
[the remark this answers is not in the document]
Comment (fhieber) [reply sockeye/train.py]: Yeah its old. I don't think we need to expose it now so I will remove the TODO.

File: sockeye/training.py [hunk 3]
```diff
@@ -844,3 +1007,77 @@ def __call__(self, epoch: int, updates: int, samples: int, tokens: int, metric:
...
+        logger.info("Starting process: %s", self.decoder_process.name)
+        self.decoder_process.start()
+
+    def collect_results(self) -> Optional[Tuple[int, Dict[str, float]]]:
+        """
+        Returns the decoded checkpoint and the decoder metrics or None if the queue is empty.
+        """
+        self.wait_to_finish()
+        if self.decoder_metric_queue.empty():
+            print("TRUE")
```
[the remark this answers is not in the document]
Comment (fhieber) [reply sockeye/training.py]: yep :)

File: sockeye/training.py [hunk 4]
```diff
@@ -14,95 +14,67 @@
...
-    :param lr_scheduler: The scheduler that lowers the learning rate during training.
-    :param gradient_compression_params: Gradient compression parameters.
+    :param gradient_compression_params: Optional dictionary of gradient compression parameters.
     """
 
     def __init__(self,
                  config: model.ModelConfig,
                  context: List[mx.context.Context],
+                 output_dir: str,
                  train_iter: data_io.BaseParallelSampleIter,
```
[the remark this answers is not in the document]
Comment (fhieber) [reply sockeye/training.py]: Its also used for the default_bucket_key and the list of buckets (see lines 130 and 139).
Do you still think we should make these 4 parameters explicit in the constructor?

[the remark this answers is not in the document]
Comment (fhieber) [reply sockeye/training.py]: I fixed this.

File: sockeye/inference.py [hunk 5]
```diff
@@ -55,34 +55,23 @@ class InferenceModel(model.SockeyeModel):
     """
 
     def __init__(self,
-                 model_folder: str,
+                 config: model.ModelConfig,
+                 params_fname: str,
```
Comment (tdomhan) [inline sockeye/inference.py]: what was wrong with optionally passing in the checkpoint? Conceptually it makes sense to me that the structure of the model folder is not exposed outside the InferenceModel.

File: sockeye/training.py [hunk 6]
```diff
@@ -146,657 +130,848 @@ def sym_gen(seq_lens):
...
-                    logger.info("Model has not improved for %d checkpoints", train_state.num_not_improved)
+                # (5) adjust learning rates
+                self._adjust_learning_rate(has_improved, lr_decay_param_reset, lr_decay_opt_states_reset)
 
-                if max_num_not_improved >= 0 and train_state.num_not_improved >= max_num_not_improved:
+                # (6) save training state
+                self._save_training_state(train_iter)
+
+                # (7) determine stopping
+                if 0 <= max_num_not_improved <= self.state.num_not_improved:
```
Comment (tdomhan) [inline sockeye/training.py]: maybe it's just me, but I find this much harder to parse than two separate >= clauses.

File: sockeye/training.py [hunk 7]
```diff
@@ -146,657 +130,848 @@ def sym_gen(seq_lens):
...
+
+        utils.check_condition(self.optimizer_config.gradient_clipping_type in C.GRADIENT_CLIPPING_TYPES,
+                              "Unknown gradient clipping type %s" % self.optimizer_config.gradient_clipping_type)
+
+        utils.check_condition(early_stopping_metric in C.METRICS,
+                              "Unsupported early-stopping metric: %s" % early_stopping_metric)
+        if early_stopping_metric == C.BLEU:
+            utils.check_condition(cp_decoder is not None, "%s requires CheckpointDecoder" % C.BLEU)
+
+    def _check_dist_kvstore_requirements(self, lr_decay_opt_states_reset:bool, lr_decay_param_reset, optimizer):
```
Comment (tdomhan) [inline sockeye/training.py]: maybe github is hiding something. Is this still called?

Comment (fhieber) [reply sockeye/training.py]: no, and its not needed anymore since this logic is in `_check_args()`. Will remove `_check_dist_kvstore_requirements`

File: sockeye/training.py (hunk 6 above)
Comment (fhieber) [reply sockeye/training.py]: I was annoyed by the IDE warning. If you feel strongly about this I can revert, otherwise I would leave as is.

File: sockeye/inference.py (hunk 5 above)
Comment (fhieber) [reply sockeye/inference.py]: There is a couple of reasons we have the `load_models()` function by now which is essentially a factory. We need it anyway so I wanted to remove as much logic from the model constructors. But if you like the checkpoint logic in there, I can move it back in.

Comment (fhieber) [reply sockeye/inference.py]: Actually, I would like to leave as is since output folder is really just needed for this logic and nowhere else in `InferenceModel`. If we pull out the param_fname logic, we need only one constructor argument (fname) instead of two (checkpoint & model_folder).

File: sockeye/training.py (hunk 6 above)
Comment (tdomhan) [reply sockeye/training.py]: sure, was just a comment, feel free to leave it like that.

File: sockeye/inference.py (hunk 5 above)
Comment (tdomhan) [reply sockeye/inference.py]: sure, it's not a big deal and I don't feel strongly about this.

File: sockeye/utils.py [hunk 8]
```diff
@@ -448,7 +449,7 @@ def get_gpu_memory_usage(ctx: List[mx.context.Context]) -> Optional[Dict[int, Tu
         ctx = [ctx]
     ctx = [c for c in ctx if c.device_type == 'gpu']
     if not ctx:
-        return None
+        return {}
```
Comment (tdomhan) [inline sockeye/utils.py]: I guess then we can drop the `Optional` from the type annotation

File: sockeye/train.py [hunk 9]
```diff
@@ -174,26 +175,26 @@ def determine_context(args: argparse.Namespace, exit_stack: ExitStack) -> List[m
...
+                              train_context: List[mx.Context]) -> Optional[checkpoint_decoder.CheckpointDecoder]:
     """
-    Determine the number of sentences to decode and the context we should run on (CPU or GPU).
+    Returns a checkpoint decoder or None.
 
     :param args: Arguments as returned by argparse.
     :param exit_stack: An ExitStack from contextlib.
     :param train_context: Context for training.
-    :return: The number of sentences to decode and a list with the context(s) to run on.
+    :return: A CheckpointDecoder if --decode-and-evaluate != 0, None else.
```
Comment (davvil) [inline sockeye/train.py]: "else None"

File: sockeye/train.py [hunk 10]
```diff
@@ -724,24 +689,42 @@ def define_optimizer(args, lr_scheduler_instance) -> Tuple[str, Dict, str, str,
...
-    logger.info("Optimizer: %s", optimizer)
-    logger.info("Optimizer Parameters: %s", optimizer_params)
-    logger.info("kvstore: %s", args.kvstore)
-    logger.info("Gradient Compression: %s", gradient_compression_params(args))
 
-    return optimizer, optimizer_params, args.kvstore, gradient_clipping_type, gradient_clipping_threshold
+    config = OptimizerConfig(name=args.optimizer,
+                             params=optimizer_params,
+                             kvstore=args.kvstore,
+                             initializer=initializer.get_initializer(
```
Comment (davvil) [inline sockeye/train.py]: I would separate the creating of the initializer and the lr_scheduler into previous calls, for clarity.

File: sockeye/train.py (hunk 9 above)
Comment (fhieber) [reply sockeye/train.py]: done

File: sockeye/train.py (hunk 10 above)
Comment (fhieber) [reply sockeye/train.py]: ok

File: sockeye/training.py [hunk 11]
```diff
@@ -146,657 +130,835 @@ def sym_gen(seq_lens):
...
+        """
+        Loads optimizer states from file.
+
+        :param fname: File name to load optimizer states from.
+        """
+        self.current_module.load_optimizer_states(fname)
+
+    def initialize_parameters(self, initializer: mx.init.Initializer, allow_missing_params: bool):
+        """
+        Initializes the parameters of the underlying module.
```
Comment (davvil) [inline sockeye/training.py]: nitpick: empty line before first :param:

File: sockeye/training.py [hunk 12]
```diff
@@ -146,657 +130,835 @@ def sym_gen(seq_lens):
...
+        """
+        arg_params, aux_params = self.module.get_params()
+        self.module.set_params(arg_params, aux_params)
+        self.params = arg_params
+        self.aux_params = aux_params
+        super().save_params_to_file(fname)
+
+    def load_params_from_file(self, fname: str):
+        """
+        Loads parameters from a file and sets the parameters of the underlying module and this model instance.
```
Comment (davvil) [inline sockeye/training.py]: No docstring for fname

File: sockeye/training.py [hunk 13]
```diff
@@ -146,657 +130,835 @@ def sym_gen(seq_lens):
...
+        """
+        super().load_params_from_file(fname)  # sets self.params & self.aux_params
+        self.module.set_params(arg_params=self.params, aux_params=self.aux_params)
+
+    def install_monitor(self, monitor_pattern: str, monitor_stat_func_name: str):
+        """
+        Installs an MXNet monitor onto the underlying module.
```
Comment (davvil) [inline sockeye/training.py]: Missing docstrings.

File: sockeye/training.py (hunk 6 above)
Comment (davvil) [reply sockeye/training.py]: FWIW I also prefer separate clauses (probably because of C backgound).

Comment (davvil) [reply sockeye/training.py]: Out of curiosity, what was the warning?

File: sockeye/utils.py [hunk 14]
```diff
@@ -32,7 +33,7 @@
 import numpy as np
 
 import sockeye.constants as C
-from sockeye import __version__
+from sockeye import __version__, constants as C
```
Comment (davvil) [inline sockeye/utils.py]: Aren't you importing constants two times?

File: sockeye/training.py (hunk 11 above)
Comment (fhieber) [reply sockeye/training.py]: fixed

File: sockeye/training.py (hunk 12 above)
Comment (fhieber) [reply sockeye/training.py]: fixed

File: sockeye/training.py (hunk 13 above)
Comment (fhieber) [reply sockeye/training.py]: fixed

File: sockeye/utils.py (hunk 14 above)
Comment (fhieber) [reply sockeye/utils.py]: yes, optimized imports
